core/src/connection/pool/task.rs

```diff
@@ -141,22 +148,24 @@ pub async fn new_for_pending_outgoing_connection<TTrans>(
 pub async fn new_for_pending_incoming_connection<TFut, TTrans>(
     connection_id: ConnectionId,
     future: TFut,
-    drop_receiver: oneshot::Receiver<Void>,
+    abort_receiver: oneshot::Receiver<PendingConnectionCommand>,
     mut events: mpsc::Sender<PendingConnectionEvent<TTrans>>,
 ) where
     TTrans: Transport,
     TFut: Future<Output = Result<TTrans::Output, TTrans::Error>> + Send + 'static,
 {
-    match futures::future::select(drop_receiver, Box::pin(future)).await {
+    match futures::future::select(abort_receiver, Box::pin(future)).await {
         Either::Left((Err(oneshot::Canceled), _)) => {
+            unreachable!("Pool never drops channel to task.");
```

Comment:
This is not quite true. It is dropped immediately after `abort` finishes because we are consuming `self` in there.

I've read through #2329 but I don't quite understand why we need to send a dedicated value now and the drop notifier isn't enough?

1. We always initialize `abort_receiver` with `Some` so there is never a case where we call `abort` and wouldn't be able to send `PendingConnectionCommand::Abort`.
1. We are still dropping the entire struct at the end of `abort` so all the data is gone anyway.
1. We are not sending any information along with `PendingConnectionCommand::Abort`.

Unless I am missing something, reacting to `Canceled` should be equivalent to what we currently have here? The `abort` method can just consume `self` and do nothing with it which will immediately drop it and trigger the drop notifier. That will still be the same public API albeit it may be worth considering removing `abort` and using `std::mem::drop` instead if we want to be explicit about the dropping.

Reply:
> This is not quite true. It is dropped immediately after abort finishes because we are consuming self in there.

Maybe there's a misunderstanding. The thing that gets dropped is `PendingConnection`, the thing that has the abort_notifier is `PendingConnectionInfo`. The `PendingConnectionInfo` stays around in the hash map of the pool.

Whether you actually need to send abort vs just taking the notfier and dropping it. I think you can do the same thing by dropping the notifier. The reason you may want the abort value here is to make the distinction between this was dropped and we aborted this. Although with the `unreachable` we're saying we would never drop this (I'm not fully sure that's true, I need to look more closely).

Reply:
> > This is not quite true. It is dropped immediately after abort finishes because we are consuming self in there.
> 
> Maybe there's a misunderstanding. The thing that gets dropped is `PendingConnection`, the thing that has the abort_notifier is `PendingConnectionInfo`. The `PendingConnectionInfo` stays around in the hash map of the pool.

Ah what a subtle difference, I missed that, thank you!

> The reason you may want the abort value here is to make the distinction between this was dropped and we aborted this. Although with the `unreachable` we're saying we would never drop this (I'm not fully sure that's true, I need to look more closely).

That is what my closer look was triggered by! We are adding a way of distinguishing the two and then asserting that one of them never happens. So why add that in the first place?

It seems to be sufficient to change the `oneshot::Sender<Void>` to an `Option<oneshot::Sender<Void>>` so we can `take` and `mem::drop` it which will trigger the cancellation. That should avoid the `unreachable` within the `select`.